src/react/tests/useQueries.test.tsx

```diff
@@ -42,4 +43,89 @@ describe('useQueries', () => {
     expect(results[1]).toMatchObject([{ data: 1 }, { data: undefined }])
     expect(results[2]).toMatchObject([{ data: 1 }, { data: 2 }])
   })
+
+  it('should keep previous data if amount of queries is the same', async () => {
+    const key1 = queryKey()
+    const key2 = queryKey()
+    const states: UseQueryResult[][] = []
+
+    function Page() {
+      const [count, setCount] = React.useState(1)
+      const result = useQueries([
+        {
+          queryKey: [key1, count],
+          keepPreviousData: true,
+          queryFn: async () => {
+            await sleep(5)
+            return count * 2
+          },
+        },
+        {
+          queryKey: [key2, count],
+          keepPreviousData: true,
+          queryFn: async () => {
+            await sleep(10)
+            return count * 5
+          },
+        },
+      ])
+      states.push(result)
+
+      React.useEffect(() => {
+        setActTimeout(() => {
+          setCount(prev => prev + 1)
+        }, 20)
+      }, [])
+
+      return null
+    }
+
+    renderWithClient(queryClient, <Page />)
+
+    await waitFor(() => expect(states.length).toBe(7))
+
+    expect(states[0]).toMatchObject([
+      {
+        status: 'loading',
+        data: undefined,
+        isPreviousData: false,
+        isFetching: true,
+      },
+      {
+        status: 'loading',
+        data: undefined,
+        isPreviousData: false,
+        isFetching: true,
+      },
+    ])
+    expect(states[1]).toMatchObject([
+      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
+      {
+        status: 'loading',
+        data: undefined,
+        isPreviousData: false,
+        isFetching: true,
+      },
+    ])
+    expect(states[2]).toMatchObject([
+      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
+      { status: 'success', data: 5, isPreviousData: false, isFetching: false },
+    ])
+    expect(states[3]).toMatchObject([
+      { status: 'success', data: 2, isPreviousData: true, isFetching: true },
+      { status: 'success', data: 5, isPreviousData: true, isFetching: true },
+    ])
+    expect(states[4]).toMatchObject([
+      { status: 'success', data: 2, isPreviousData: true, isFetching: true },
+      { status: 'success', data: 5, isPreviousData: true, isFetching: true },
+    ])
+    expect(states[5]).toMatchObject([
+      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
+      { status: 'success', data: 5, isPreviousData: true, isFetching: true },
+    ])
+    expect(states[6]).toMatchObject([
+      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
+      { status: 'success', data: 10, isPreviousData: false, isFetching: false },
+    ])
+  })
```

Comment:
I created two more test cases, tell me what you think: 

```js
  it('should keep previous data if amount of queries is the same when has 3 elements', async () => {
    const key1 = queryKey()
    const key2 = queryKey()
    const key3 = queryKey()
    const states: UseQueryResult[][] = []

    function Page() {
      const [count, setCount] = React.useState(1)

      const result = useQueries([
        {
          queryKey: [key1, count],
          keepPreviousData: true,
          queryFn: async () => {
            await sleep(5)
            return 1 * count
          },
        },
        {
          queryKey: [key2, count],
          keepPreviousData: true,
          queryFn: async () => {
            await sleep(10)
            return 2 * count
          },
        },
        {
          queryKey: [key3, count],
          keepPreviousData: true,
          queryFn: async () => {
            await sleep(15)
            return 3 * count
          },
        },
      ])
      states.push(result)

      React.useEffect(() => {
        setActTimeout(() => {
          setCount(prev => prev + 1)
        }, 20)
      }, [])

      return null
    }

    renderWithClient(queryClient, <Page />)

    await waitFor(() => expect(states.length).toBe(9))

    expect(states[0]).toMatchObject([
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[1]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[2]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[3]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 3, isPreviousData: false, isFetching: false },
    ])

    expect(states[4]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: true, isFetching: true },
      { status: 'success', data: 2, isPreviousData: true, isFetching: true },
      { status: 'success', data: 3, isPreviousData: true, isFetching: true },
    ])

    expect(states[5]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: true, isFetching: true },
      { status: 'success', data: 2, isPreviousData: true, isFetching: true },
      { status: 'success', data: 3, isPreviousData: true, isFetching: true },
    ])

    expect(states[6]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 2, isPreviousData: true, isFetching: true },
      { status: 'success', data: 3, isPreviousData: true, isFetching: true },
    ])

    expect(states[7]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
      { status: 'success', data: 3, isPreviousData: true, isFetching: true },
    ])

    expect(states[8]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
      { status: 'success', data: 6, isPreviousData: false, isFetching: false },
    ])
  })

  it('should not keep previous data', async () => {
    const key1 = queryKey()
    const key2 = queryKey()
    const key3 = queryKey()
    const states: UseQueryResult[][] = []

    function Page() {
      const [count, setCount] = React.useState(1)

      const result = useQueries([
        {
          queryKey: [key1, count],
          queryFn: async () => {
            await sleep(5)
            return 1 * count
          },
        },
        {
          queryKey: [key2, count],
          queryFn: async () => {
            await sleep(10)
            return 2 * count
          },
        },
        {
          queryKey: [key3, count],
          queryFn: async () => {
            await sleep(15)
            return 3 * count
          },
        },
      ])
      states.push(result)

      React.useEffect(() => {
        setActTimeout(() => {
          setCount(prev => prev + 1)
        }, 20)
      }, [])

      return null
    }

    renderWithClient(queryClient, <Page />)

    await waitFor(() => expect(states.length).toBe(9))

    expect(states[0]).toMatchObject([
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[1]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[2]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[3]).toMatchObject([
      { status: 'success', data: 1, isPreviousData: false, isFetching: false },
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 3, isPreviousData: false, isFetching: false },
    ])

    expect(states[4]).toMatchObject([
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[5]).toMatchObject([
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[6]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[7]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
      {
        status: 'loading',
        data: undefined,
        isPreviousData: false,
        isFetching: true,
      },
    ])

    expect(states[8]).toMatchObject([
      { status: 'success', data: 2, isPreviousData: false, isFetching: false },
      { status: 'success', data: 4, isPreviousData: false, isFetching: false },
      { status: 'success', data: 6, isPreviousData: false, isFetching: false },
    ])
  })


```

Reply:
>  when has 3 elements

thanks, but your test has 3 static elements. My claim is that it might not work with _dynamic_ elements, so 2 elements on the first render, followed by 3 elements on the next render. I added a test for this here: 62ce4ff and it seems to work okay.

I'm still positive that if you remove one element in the middle of the list, you will keep the previous data from this item. not sure if this will be perceived as intended behaviour, which is why I've suggested a different solution here: https://github.com/tannerlinsley/react-query/pull/2340#discussion_r659193661, which would limit keepPreviousData to arrays of the same length...